### ISSUE TITLE
small externally stored row wastes memory

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#2302]
This is a managed_vector<atomic_cell> for a 10-column schema created by `cassandra-stress write -col 'n=FIXED(10)'`

    _size = 10, 
    _capacity = 13, 
    _data = 0x6000021c6368

The schema has 10 columns, why to we reserve 13?

**Comment by pdziepak** [comment]
Well, managed_vector behaves like std::vector and grows exponentially to ensure amortised O(1) insertion.  The obvious drawback is wasted space especially when the vector has more capacity than there are columns.

**Comment by avikivity** [comment]
There are reserve() and shrink_to_fit() for that

**Comment by pdziepak** [comment]
True, `reserve()` would be best. If I remember correctly it wasn't trivial to use reserve with the way cql layer creates mutation (poor excuse, but a reason why it wasn't done when managed_vector was introduced), `shrink_to_fit()` will cause a reallocation, but perhaps that's not a big deal considering that we already copy data several times.

IMR forces everyone to use two phase serialisation (first phase to determine size and the second to actually write data), so cql layer will have to be accommodated anyway.

**Comment by xemul** [comment]
The row's storage is `managed_vector<cell_and_hash, 5 /* inline size */, 32 /* maximum size */>`. Also it's `maybe_grow` method has a wierd formula for new capacity, but it will never be less than (SUDDENLY) 13 and then grows doubling.

The radix tree in its current incarnation doesn't have inline storage, and allocates nodes for 4, 16 and 48 cells.

So the "unused memory for cells" until the vector is switched to rbtree would looks like

````
cells vector tree
1      4       3
2      3       2
3      2       1
4      1       0
5      0       11    <- tree loses here and ...
6      13      10
7      12      9
8      11      8
9      10      7
10     9       6
11     8       5
12     7       4
13     6       3
14     17      2
15     16      1
16     15      0
17     14      31    <-\
...                     |- ... in his area
26     5       22    <-/
27     30      21
...
32     25      16
````
